### PR TITLE
DOC: Document the internal workings of PY_ARRAY_UNIQUE_SYMBOL

### LINKED ISSUE #9309: DOC: document PY_ARRAY_UNIQUE_SYMBOL and friends
There doesn't seem to be anywhere in the docs (or anywhere on the internet) where the role that `PY_ARRAY_UNIQUE_SYMBOL` and `NO_IMPORT` or `NO_IMPORT_ARRAY` play in creating C extensions that use the NumPy API in more than one compilation unit. We should add it!

### PR DESCRIPTION
**Pull request by jaimefrio**
Closes #9309.

### CODE REVIEW

File: doc/source/reference/c-api.array.rst [hunk 1]
```diff
@@ -3063,6 +3063,23 @@ the C-API is needed then some additional steps must be taken.
     header file as long as you make sure that NO_IMPORT_ARRAY is
     #defined before #including that file.
 
+    Internally, these #defines work as follows:
+
+        * If neither is defined, the C-API is declared to be ``static void**``,
```
Comment (charris) [inline doc/source/reference/c-api.array.rst]: Should these be ``` :c:type: `static void**` ```

File: doc/source/reference/c-api.array.rst [hunk 2]
```diff
@@ -3063,6 +3063,23 @@ the C-API is needed then some additional steps must be taken.
...
+          :c:macro:`NO_IMPORT_ARRAY` is not, the C-API is declared to be
+          ``void**``, so that it will also be visible to other compilation
+          units.
+        * If :c:macro:`NO_IMPORT_ARRAY` is #defined, regardless of whether
+          :c:macro:`PY_ARRAY_UNIQUE_SYMBOL` is, the C-API is declared to be
+          ``extern void**``, so it is expected to be defined in another
+          compilation unit.
+        * Whenever :c:macro:`PY_ARRAY_UNIQUE_SYMBOL` is #defined, it also
+          changes the name of the variable holding the C-API, which defaults
+          to ``PyArray_API``, to whatever the macro is #defined to.
```
Comment (charris) [inline doc/source/reference/c-api.array.rst]: ``` :c:data: `PyArray_API` ```?

**Comment by jaimefrio** [comment]
Have added your suggested tags, plus have reformatted the text to fit in 72 columns.

**Comment by charris** [comment]
Thanks Jaime.

**Pull request merged**